server/cmwell-cons/src/main/scala/ctl.scala

```diff
@@ -2174,6 +2180,25 @@ abstract class Host(user: String,
 
     info("  updating version history")
     dataInitializer.logVersionUpgrade(hosts(0))
+
+    val upgradedVersion = extractVersionFromCmwellProperties
+    info(s"Upgrading to version: $upgradedVersion")
+
+    Upgrade.runPostUpgradeActions(currentVersion, upgradedVersion)
+    info(s"Upgrade completed!")
+  }
+
+  def extractVersionFromCmwellProperties : String = {
+    val cmwellProp = Source.fromURL(this.getClass.getResource(cmwellPropertiesFile)).mkString
+    (Json.parse(cmwellProp) \ "cm-well_version").as[String].replace("x-SNAPSHOT", "0")
+  }
+
+  def extractVersionFromProcNode(host : String) : Future[String] = {
```

Comment:
You can wrap the entire function body with `cmwell.util.concurrent.retry(...) { ... }`